p2p/protocol/autonatv2/client.go

```diff
@@ -0,0 +1,308 @@
+	}
+
+	return Result{
+		Idx:          idx,
+		Addr:         addr,
+		Reachability: rch,
+		Status:       resp.DialStatus,
+	}, nil
+}
+
+func (ac *client) sendDialData(req *pb.DialDataRequest, w pbio.Writer, msg *pb.Message) error {
+	nb := req.GetNumBytes()
+	ddResp := &pb.DialDataResponse{Data: ac.dialData}
+	*msg = pb.Message{
+		Msg: &pb.Message_DialDataResponse{
+			DialDataResponse: ddResp,
+		},
+	}
+	for remain := int(nb); remain > 0; {
+		end := remain
+		if end > len(ddResp.Data) {
+			end = len(ddResp.Data)
+		}
+		ddResp.Data = ddResp.Data[:end]
+		if err := w.WriteMsg(msg); err != nil {
+			return err
+		}
+		remain -= end
+	}
+	return nil
+}
+
+func newDialRequest(reqs []Request, nonce uint64) pb.Message {
+	addrbs := make([][]byte, len(reqs))
+	for i, r := range reqs {
+		addrbs[i] = r.Addr.Bytes()
+	}
+	return pb.Message{
+		Msg: &pb.Message_DialRequest{
+			DialRequest: &pb.DialRequest{
+				Addrs: addrbs,
+				Nonce: nonce,
+			},
+		},
+	}
+}
+
+// handleDialBack receives the nonce on the dial-back stream
+func (ac *client) handleDialBack(s network.Stream) {
+	if err := s.Scope().SetService(ServiceName); err != nil {
+		log.Debugf("failed to attach stream to service %s: %w", ServiceName, err)
+		s.Reset()
+		return
+	}
+
+	if err := s.Scope().ReserveMemory(dialBackMaxMsgSize, network.ReservationPriorityAlways); err != nil {
+		log.Debugf("failed to reserve memory for stream %s: %w", DialBackProtocol, err)
+		s.Reset()
+		return
+	}
+	defer s.Scope().ReleaseMemory(dialBackMaxMsgSize)
+
+	s.SetDeadline(time.Now().Add(dialBackStreamTimeout))
+	defer s.Close()
+
+	r := pbio.NewDelimitedReader(s, dialBackMaxMsgSize)
+	var msg pb.DialBack
+	if err := r.ReadMsg(&msg); err != nil {
+		log.Debugf("failed to read dialback msg from %s: %s", s.Conn().RemotePeer(), err)
+		s.Reset()
+		return
+	}
+	nonce := msg.GetNonce()
+
+	ac.mu.Lock()
+	ch := ac.dialBackQueues[nonce]
+	ac.mu.Unlock()
+	if ch == nil {
+		log.Debugf("dialback received with invalid nonce: localAdds: %s peer: %s nonce: %d", s.Conn().LocalMultiaddr(), s.Conn().RemotePeer(), nonce)
+		s.Reset()
+		return
+	}
+	select {
+	case ch <- s.Conn().LocalMultiaddr():
+	default:
+		log.Debugf("multiple dialbacks received: localAddr: %s peer: %s", s.Conn().LocalMultiaddr(), s.Conn().RemotePeer())
+		s.Reset()
+		return
+	}
+	w := pbio.NewDelimitedWriter(s)
+	res := pb.DialBackResponse{}
+	if err := w.WriteMsg(&res); err != nil {
+		log.Debugf("failed to write dialback response: %s", err)
+		s.Reset()
+	}
+}
+
+func areAddrsConsistent(local, external ma.Multiaddr) bool {
+	if local == nil || external == nil {
+		return false
+	}
+
+	localProtos := local.Protocols()
+	externalProtos := external.Protocols()
+	if len(localProtos) != len(externalProtos) {
+		return false
+	}
+	for i := 0; i < len(localProtos); i++ {
+		if i == 0 {
+			if localProtos[i].Code == externalProtos[i].Code ||
+				localProtos[i].Code == ma.P_IP6 && externalProtos[i].Code == ma.P_IP4 /* NAT64 */ {
```

Comment:
I don't understand what you mean here. Can you elaborate?

Reply:
I will remove this but this was to support NAT64 addresses. Locally they appear to be IPv6 and remotely they are IPv4. But we can add that support later when someone asks for it. I'm not sure if rest of the code works with NAT64 anyway.